fix to sqlite_recorder so only root records to db

fzahle opened the pull request:
I can't run in parallel on the latest master with the changes made to SqliteRecorder. It looks like only the root proc returns from the record calls and the other procs raise this warning and cause the process to stall:

    TypeError: 'NoneType' object does not support item assignment

since they all set self.out to `None`.

adding a return statement in `record_metadata` and `record_iteration` seems to fix it so that only the root tries to record.

hschilling commented:
Frederik, thanks for the pull request and your support in general!

 At least for the codes and tests we wrote, this added code doesn't seem to be needed, although it does not hurt. When running in parallel, with a parallel recorder like sqlite_recorder, the RecordingManager should only call record_metadata and record_iteration when on rank 0. I just ran, using the latest master, a parallel optimization test while recording to sqlite and also put in some print statements to see when record_iteration and record_metadata are called. They are only called on rank 0. Maybe I am missing something or maybe you are doing something more complicated than we are? Can you tell us what you are doing so I can try it here? 

Thanks

fzahle commented:
You're right. So the issue seems to be in `RecordingManager`. I can see that `root.comm.rank` does not report the same as `MPI.COMM_WORLD.rank` on my Mac. So when I change the if statements in `record_iteration` and `record_metadata` in `RecordingManager`, it runs. So either there's an issue with my MPI install or some issue with how `root.comm.rank` is set.

fzahle commented:
forgot to explain what the setup is: I'm running a simple optimisation with a parallel FD group as the root. so is it because the parallel FD groups don't get the correct comm's passed to them? I'll try to put together a smaller test problem to narrow it down.

this PR clearly doesn't fix the problem in the right place so should probably be closed.

fzahle commented:
The issue seems to be that when MPI.COMM_WORLD.Split(<int>) is called in `ParallelFDGroup._setup_communicators` it creates a new world where it will then be rank zero. This small example shows that:

    from mpi4py import MPI
    subcomm = MPI.COMM_WORLD.Split(MPI.COMM_WORLD.rank)
    print('subcomm rank: %i, %i'%(subcomm.rank, MPI.COMM_WORLD.rank))

I haven't come across issues in any other places than in the recorder, but I guess you should use `MPI.COMM_WORLD.rank` everywhere in the code rather than `comm.rank` to get the "physical" rank. I don't know whether you think a fix in `RecordingManager` is OK or it can be fixed closer to the "root"?

hschilling commented:
Frederik,

 So are you saying you changed the 

            if recorder._parallel or root.comm.rank == 0:

to be 

            if recorder._parallel or MPI.COMM_WORLD.rank == 0:


?

in record_metadata and record_iteration?

Thanks

fzahle commented:
that was what I was about to propose, but that only works in parallel, so you'd have to split that if statement up and first check `if MPI` I guess.

fzahle commented:
this second commit fixes the issue in the recording manager, but maybe this could have some unintentional side effects when not checking the root's rank?

hschilling commented:
Thanks. That commit looks good to me!
